src/compas_timber/connections/joint.py

```diff
@@ -58,13 +58,37 @@ class Joint(Interaction):
     """
 
     SUPPORTED_TOPOLOGY = JointTopology.TOPO_UNKNOWN
+    MIN_ELEMENT_COUNT = 2
+    MAX_ELEMENT_COUNT = 2
 
     def __init__(self, **kwargs):
         super(Joint, self).__init__(name=self.__class__.__name__)
+        self.elements = []
 
     @property
     def beams(self):
-        raise NotImplementedError
+        for element in self.elements:
+            if getattr(element, "is_beam", False):
+                yield element
+
+    @property
+    def plates(self):
+        for element in self.elements:
+            if getattr(element, "is_plate", False):
+                yield element
+
+    @property
+    def fasteners(self):
+        for element in self.elements:
+            if getattr(element, "is_fastener", False):
+                yield element
```

Comment:
I'm not convinced about this.
my concern is that any introduction of a new element type would involve changes in many places.

fastener joints can simply have a fastener attribute (and probably already have) and in any case any arbitrary queries can be done on the model.

another concern is having these properties become more and more ubiquitous around the library and used rather casually in many places and end up incurring hard to spot performance cost.

Reply:
This is already how it's done at the `TimberModel` level. I asked about this for a previous PR, whether we can have the element type register itself with the Joint class, instead of having to add it manually.